network-manager-app: Add support for OpenThread BR integration on OpenWrt

ksperling-apple opened the pull request:
### Summary

Add a new ThreadBROpenThreadUbus implementation of the TBRM delegate that interacts with otbr-agent via ubus. Rename the existing dummy implementation to ThreadBRFake.h.

Add some generic utilities for interacting with ubus (and the underlying uloop / ubox libraries) from C++. At the moment this lives under the network-manager example directory, but could move into platform/Linux in the core SDK at a later stage. (Future work based on this is likely to include interacting with hostapd via ubus, as well as exposing some information about the matter application itself, e.g. pairing status.)

### Testing

Tested manually on OpenWrt for now, since this integration code is not easy to meaningfully test in unit tests.

bzbarsky-apple commented on src/system/SystemLayerImplSelect.cpp:
```diff
@@ -685,16 +695,21 @@ void LayerImplSelect::HandleEvents()
     for (auto & w : mSocketWatchPool)
     {
-        if (w.mFD != kInvalidFd)
+        if (w.mFD != kInvalidFd && w.mCallback != nullptr)
```
Hmm.  Wouldn't we want to drain the events even if there is no callback?

ksperling-apple replied:
`SocketEventsFromFDs` doesn't drain anything, it just sets a bit mask based on which of the socket sets then given FD appears in. So this just avoids a bit of unnecessary work.

bzbarsky-apple commented on src/lib/support/IntrusiveList.h:
Might have been nice to have these various infrastructure changes in separate PRs with clear descriptions...

ksperling-apple replied:
Done as #33968 (with tests!)

bzbarsky-apple commented on examples/network-manager-app/linux/BUILD.gn:
```diff
@@ -15,7 +15,11 @@
 
-executable("network-manager-app") {
+declare_args() {
+  matter_enable_ubus = false
```
Please document what this arg means.

bzbarsky-apple commented on examples/network-manager-app/linux/UbusManager.cpp:
```diff
@@ -0,0 +1,80 @@
+    UloopHandler::Register();
+
+    int status;
+    if ((status = ubus_connect_ctx(&mContext, nullptr)))
```
My knowledge of ubus is nonexistent, so I am just rubber-stamping the ubus bits.  If you need me to actually go read up on these APIs and do a proper review, please let me know.

ksperling-apple commented:
> > * Add the ability for EventLoopHandlers to participate in the event loop
> > * Update network-manager-app with correct device type and integrate ubus
> 
> Could we split this into separate PRs? Support for event loop handlers seem reasonably independent from the network manager changes.
> 
> Can we add some unit tests that validates that EventLoopHandlers work as expected?

I will split it out into a separate PR

andy31415 commented:
@ksperling-apple this is quite an old PR and has approvals, but also merge conflicts. Could you review if it is appropriate to update or if it should be closed if not applicable anymore?

ksperling-apple commented:
> @ksperling-apple this is quite an old PR and has approvals, but also merge conflicts. Could you review if it is appropriate to update or if it should be closed if not applicable anymore?

Planning to pick this back up in the near future.

ksperling-apple committed:
lib/support: Add CompileTimeString

ksperling-apple committed:
lib/support: Document and test in-place HexToBytes

ksperling-apple committed:
SystemLayerSelect: Don't log EINTR as an error

ksperling-apple committed:
network-manager-app: Add support for OpenThread BR integration on OpenWrt

ksperling-apple committed:
Address review comments / tweaks

- Use SA_RESETHAND instead of manual resetting to SIG_DFL
- Call ubus_shutdown() in HandleConnectionLost
- Re-install connection_lost handler in Connect()

ksperling-apple commented on examples/network-manager-app/linux/ThreadBROpenThreadUbus.cpp:
```diff
@@ -0,0 +1,160 @@
+uint16_t OpenThreadUbusBorderRouterDelegate::GetThreadVersion()
+{
+    return /* Thread 1.4.0 */ 5;
+}
```
The existing comment is sufficient.

ksperling-apple commented on examples/network-manager-app/linux/UloopHandler.cpp:
```diff
@@ -0,0 +1,183 @@
+    guard.sa_handler = [](int caught) {
+        signal(caught, SIG_DFL); // restore SIG_DFL (don't worry about masks etc)
+        raise(caught);           // re-raise the signal to the default handler
+    };
```
Addressed by using SA_RESETHAND instead of manual resetting.

ksperling-apple committed:
Comment wording

ksperling-apple committed:
Enable the dummy implementation of Thread Network Diagnostics server by default
